lib/puppet/util/execution.rb

```diff
@@ -216,28 +228,32 @@ def self.execute(command, options = NoOptionsSpecified)
       elsif Puppet.features.microsoft_windows?
         process_info = execute_windows(*exec_args)
         begin
+          [stdin, stdout, stderr].each {|io| io.close rescue nil}
+          unless options[:squelch]
+            while !reader.eof?
+              output << reader.read
+            end
+            read_succeeded = true
+          end
           exit_status = Puppet::Util::Windows::Process.wait_process(process_info.process_handle)
         ensure
           FFI::WIN32.CloseHandle(process_info.process_handle)
           FFI::WIN32.CloseHandle(process_info.thread_handle)
         end
       end
 
-      [stdin, stdout, stderr].each {|io| io.close rescue nil}
-
-      # read output in if required
-      unless options[:squelch]
-        output = wait_for_output(stdout)
-        Puppet.warning "Could not get output" unless output
+      unless options[:squelch] || read_succeeded
+        Puppet.warning "Could not get output"
+        output = nil
       end
 
       if options[:failonfail] and exit_status != 0
         raise Puppet::ExecutionFailure, "Execution of '#{command_str}' returned #{exit_status}: #{output.strip}"
       end
     ensure
-      if !options[:squelch] && stdout
-        # if we opened a temp file for stdout, we need to clean it up.
-        stdout.close!
+      if !options[:squelch] && reader
+        # if we opened a pipe, we need to clean it up.
+        reader.close
```

Comment:
That does seem like a loose end. However, as far as I can tell if we never fork and then close one end, the other end also closes. So in that scenario the pipe would still be closed.

I'm not sure if that leaves dangling handles for the files. We could still attempt to close all handles in the ensure block anyway.